makelib/golang.mk

```diff
@@ -333,22 +368,27 @@ $(GOFMT):
 	@$(OK) installing gofmt$(GOFMT_VERSION)
 
 $(GOIMPORTS):
-	@$(INFO) installing goimports
-	@mkdir -p $(TOOLS_HOST_DIR)/tmp-imports || $(FAIL)
-	@GOPATH=$(TOOLS_HOST_DIR)/tmp-imports GOBIN=$(TOOLS_HOST_DIR) $(GOHOST) get -u golang.org/x/tools/cmd/goimports || rm -fr $(TOOLS_HOST_DIR)/tmp-imports || $(FAIL)
-	@rm -fr $(TOOLS_HOST_DIR)/tmp-imports
-	@$(OK) installing goimports
+	@$(MAKE) go.get GOPKG=golang.org/x/tools/cmd/goimports
 
 $(GOJUNIT):
-	@$(INFO) installing go-junit-report
-	@mkdir -p $(TOOLS_HOST_DIR)/tmp-junit || $(FAIL)
-	@GOPATH=$(TOOLS_HOST_DIR)/tmp-junit GOBIN=$(TOOLS_HOST_DIR) $(GOHOST) get github.com/jstemmer/go-junit-report || rm -fr $(TOOLS_HOST_DIR)/tmp-junit || $(FAIL)
-	@rm -fr $(TOOLS_HOST_DIR)/tmp-junit
-	@$(OK) installing go-junit-report
-
-$(GOCOVER_COBERTURA):
-	@$(INFO) installing gocover-cobertura
-	@mkdir -p $(TOOLS_HOST_DIR)/tmp-gocover-cobertura || $(FAIL)
-	@GOPATH=$(TOOLS_HOST_DIR)/tmp-gocover-cobertura GOBIN=$(TOOLS_HOST_DIR) $(GOHOST) get github.com/t-yuki/gocover-cobertura || rm -fr $(TOOLS_HOST_DIR)/tmp-covcover-cobertura || $(FAIL)
-	@rm -fr $(TOOLS_HOST_DIR)/tmp-gocover-cobertura
-	@$(OK) installing gocover-cobertura
+	@$(MAKE) go.get GOPKG=github.com/jstemmer/go-junit-report
+
+$(GOCOVER_COBERTURA): $(GOJUNIT)
+	@$(MAKE) go.get GOPKG=github.com/t-yuki/gocover-cobertura
+
+go.get:
+	@$(INFO) installing $(GOPKG)
+	$(eval GO_GET_GOPATH := $(GOPATH))
+ifneq ($(GO111MODULE),on)
+	$(eval GO_GET_GOPATH := $(TOOLS_HOST_DIR)/tmp)
+	@mkdir -p $(GO_GET_GOPATH) || $(FAIL)
+endif
+	@GOPATH=$(GO_GET_GOPATH) GOBIN=$(TOOLS_HOST_DIR) $(GOHOST) get $(GOPKG) || $(FAIL)
+ifneq ($(GO111MODULE),on)
+	@rm -rf $(GO_GET_GOPATH)
+else
+	@# the below command reverts the go.mod modification because of running `go get`
+	@# otherwise the build becomes dirty and won't be published
+	@$(GOHOST) mod tidy
```

Comment:
is this needed when `GOPATH` is going to a temp directory?

Reply:
As long as  the project directory where `go.mod` resides is a parent to `GOPATH`, running `go get PKG` command  adds that pkg definition to `go.mod`.

Note that since the targets and tools in build repository are supposed to be independent from consuming repositories `go.mod`, it shouldn't be assumed that these packages are already part of `go.mod`

Reply:
I'm still not following why if we are using a unique `GOPATH` why we need to tidy anything. maybe you can find me today and discuss.

Reply:
per our discussion, added temporary `go.mod` file in the tmp folder to avoid this